java/arcs/android/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,210 @@
+ *         myHelper.onStartCommand(intent)
+ *         return result
+ *     }
+ * }
+ * ```
+ */
+class ArcHostHelper(private val service: Service, private val arcHost: ArcHost) {
+
+    private val job = Job() + Dispatchers.IO + CoroutineName("ArcHostHelper")
+
+    /**
+     * Determines whether or not the given [Intent] represents a call to an [ArcHost] and invokes
+     * the appropriate interface methods.
+     */
+    fun onStartCommand(intent: Intent?) = CoroutineScope(job).launch {
+        onStartCommandSuspendable(intent)
+    }
+
+    // VisibleForTesting
+    suspend fun onStartCommandSuspendable(intent: Intent?) {
+        // Ignore other actions
+        if (intent?.action?.startsWith(ARC_HOST_INTENT) != true) return
+
+        // Ignore Intent when it doesn't target our Service
+        if (intent.component?.equals(ComponentName(service, service::class.java)) != true) return
+
+        when (intent.getSerializableExtra(OPERATION) as Operation?) {
+            Operation.START_ARC -> runWithResult(intent) {
+                arcHost.startArc(
+                    intent.getParcelableExtra<ParcelablePlanPartition?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.STOP_ARC -> runWithResult(intent) {
+                arcHost.stopArc(
+                    intent.getParcelableExtra<ParcelablePlanPartition?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.REGISTER_PARTICLE -> runWithResult(intent) {
+                arcHost.registerParticle(
+                    intent.getParcelableExtra<ParcelableParticleIdentifier?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.UNREGISTER_PARTICLE -> runWithResult(intent) {
+                arcHost.unregisterParticle(
+                    intent.getParcelableExtra<ParcelableParticleIdentifier?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.GET_REGISTERED_PARTICLES -> runWithResult(intent) {
+                arcHost.registeredParticles()
+            }
+            else -> {
+                throw IllegalArgumentException("Operation $intent not implemented.")
+            }
+        }
+    }
+
+    /**
+     * Execute an [ArcHost] method in the block, encode return into a [Bundle] and send via
+     * [ResultReceiver]
+     */
+    private suspend fun runWithResult(intent: Intent, block: suspend () -> Any?) {
+        val bundle = Bundle()
+        val result = block()
+
+        when (result) {
+            is List<*> -> {
+                val arrayList = ArrayList<ParcelableParticleIdentifier>()
+                result.forEach {
+                    it as ParticleIdentifier
+                    arrayList += it.toParcelable()
```

Comment:
Optional: I think this would be nicer inlined as `(it as ParticleIdentifier).toParcelable()`